internal/cache/persistent_lru_cache.go

```diff
@@ -295,86 +233,92 @@ func newContentMetadataHeap() contentMetadataHeap {
+
+	h.index[bm.BlobID] = len(h.data)
+	h.data = append(h.data, bm)
+	h.totalDataBytes += bm.Length
+}
+
+func (h *contentMetadataHeap) AddOrUpdate(bm blob.Metadata) {
 	if i, exists := h.index[bm.BlobID]; exists {
 		// only accept newer timestamps
-		if h.data[i].metadata.Timestamp.IsZero() || bm.Timestamp.After(h.data[i].metadata.Timestamp) {
-			h.dataSize += bm.Length - h.data[i].metadata.Length
-			h.data[i] = &blobCacheEntry{metadata: bm}
+		if bm.Timestamp.After(h.data[i].Timestamp) {
+			h.totalDataBytes += bm.Length - h.data[i].Length
+			h.data[i] = bm
 			heap.Fix(h, i)
 		}
 	} else {
-		h.index[bm.BlobID] = len(h.data)
-		h.data = append(h.data, &blobCacheEntry{metadata: bm})
-		h.dataSize += bm.Length
+		heap.Push(h, bm)
 	}
 }
 
-func (h *contentMetadataHeap) Pop() interface{} {
+func (h *contentMetadataHeap) Pop() any {
 	old := h.data
 	n := len(old)
 	item := old[n-1]
 	h.data = old[0 : n-1]
-	h.dataSize -= item.metadata.Length
-	delete(h.index, item.metadata.BlobID)
+	h.totalDataBytes -= item.Length
+	delete(h.index, item.BlobID)
 
-	return item.metadata
+	return item
+}
+
+// +checklocks:c.listCacheMutex
+func (c *PersistentCache) aboveSoftLimit(extraBytes int64) bool {
+	return c.listCache.totalDataBytes+extraBytes+c.pendingWriteBytes > c.sweep.MaxSizeBytes
 }
 
-func (h *contentMetadataHeap) LookupByID(id blob.ID) (int, *blobCacheEntry) {
-	i, ok := h.index[id]
-	if !ok {
-		return -1, nil
+// +checklocks:c.listCacheMutex
+func (c *PersistentCache) aboveHardLimit(extraBytes int64) bool {
+	if c.sweep.LimitBytes <= 0 {
+		return false
 	}
 
-	return i, h.data[i]
+	return c.listCache.totalDataBytes+extraBytes+c.pendingWriteBytes > c.sweep.LimitBytes
 }
 
-func (h contentMetadataHeap) DataSize() int64 { return h.dataSize }
-
 // +checklocks:c.listCacheMutex
-func (c *PersistentCache) listCacheCleanupLocked(ctx context.Context) {
+func (c *PersistentCache) sweepLocked(ctx context.Context) {
 	var (
 		unsuccessfulDeletes     []blob.Metadata
-		unsuccessfulDeletesSize int64
+		unsuccessfulDeleteBytes int64
 		now                     = c.timeNow()
 	)
 
-	// if there are blobs pending to be deleted ...
-	for c.listCache.DataSize() > 0 &&
-		// ... and everything including what we couldn't delete is still bigger than the threshold
-		(c.listCache.DataSize()+unsuccessfulDeletesSize) > c.sweep.MaxSizeBytes {
-		oldest := heap.Pop(&c.listCache).(blob.Metadata) //nolint:forcetypeassert
+	for len(c.listCache.data) > 0 && (c.aboveSoftLimit(unsuccessfulDeleteBytes) || c.aboveHardLimit(unsuccessfulDeleteBytes)) {
+		// examine the oldest cache item without removing it from the heap.
+		oldest := c.listCache.data[0]
 
-		// stop here if the oldest item is below the specified minimal age
-		if age := now.Sub(oldest.Timestamp); age < c.sweep.MinSweepAge {
-			heap.Push(&c.listCache, oldest)
+		if age := now.Sub(oldest.Timestamp); age < c.sweep.MinSweepAge && !c.aboveHardLimit(unsuccessfulDeleteBytes) {
```

Comment:
nothing to change here; just leaving a quick note about possible cache attrition

So, with this PR we'll override the minimum age rule if we are above hard-limit. This helps maintaining the cache hard limit invariant at all times and simplify things.

I am not 100% sure yet which is a better override ? I mean, overriding min-age vs overriding the hard-limit and allowing blobs with the min-age to decay in timely fashion. Could overriding min-age lead to cache attrition ? Do you know the original intent of having a min-age limit for the kopia cache ?